### PR TITLE
[WIP] check for inverted axis and custom CRS

### PR DESCRIPTION
**Pull request by vincentsarago**
Note: This PR will need a new rasterio release to be able to work

closes #8 #26

### CODE REVIEW

File: morecantile/models.py [hunk 1]
```diff
@@ -57,16 +76,34 @@ class TileMatrixSet(BaseModel):
     abstract: Optional[str]
     keywords: Optional[List[str]]
     identifier: str = Field(..., regex=r"^[\w\d_\-]+$")
-    supportedCRS: str = Field(..., regex=r"^http://www.opengis.net/")
+    supportedCRS: Union[CRS, AnyHttpUrl]
```
Comment (vincentsarago) [inline morecantile/models.py:79]: We now store the CRS directly if passed

File: morecantile/models.py [hunk 2]
```diff
@@ -57,16 +76,34 @@ class TileMatrixSet(BaseModel):
...
     wellKnownScaleSet: Optional[AnyHttpUrl] = None
     boundingBox: Optional[BoundingBox]
     tileMatrix: List[TileMatrix]
 
+    class Config:
+        """Configure TileMatrixSet."""
+
+        arbitrary_types_allowed = True
+        json_encoders = {
+            CRS: lambda v: CRS_to_uri(v),
```
Comment (vincentsarago) [inline morecantile/models.py:89]: when returning a JSON representation of the model we will create an URI

File: morecantile/models.py [hunk 3]
```diff
@@ -449,12 +502,14 @@ def tiles(
...
-            w = max(tms_bounds[0], w)
-            s = max(tms_bounds[1], s)
-            e = min(tms_bounds[2], e)
-            n = min(tms_bounds[3], n)
+            if not self.intersect_tms(CoordsBbox(w, s, e, n)):
+                continue
+            # w = max(tms_bounds[0], w)
+            # s = max(tms_bounds[1], s)
+            # e = min(tms_bounds[2], e)
+            # n = min(tms_bounds[3], n)
```
Comment (vincentsarago) [inline morecantile/models.py:512]: WIP

File: setup.py [hunk 4]
```diff
@@ -5,7 +5,7 @@
 with open("README.md") as f:
     long_description = f.read()
 
-inst_reqs = ["rasterio>=1.1", "pydantic"]
+inst_reqs = ["rasterio>=1.1.6", "pydantic"]
```
Comment (vincentsarago) [inline setup.py:8]: TBD

File: CHANGES.txt [hunk 5]
```diff
@@ -1,3 +1,7 @@
+1.2.1 (2020-06-02)
+------------------
+- Truncate `morecantile.TileMatrixSet.tiles` to the maximum TMS bounds.
+
```
Comment (vincentsarago) [inline CHANGES.txt:4]: Need to change

**Comment by vincentsarago** [comment]
FYI, this is a sketch I need to figure out private variable in pydantic

**Comment by vincentsarago** [comment]
With latest commit 
```python
import morecantile
nztm = morecantile.tms.get('NZTM2000')                                                                                                                                                                                                                                                                    
print(nztm.xy_bounds(morecantile.Tile(1, 2, 0)))                                                                                                                                                                                                                                                          
CoordsBbox(xmin=1293760.0, ymin=3118720.0, xmax=3587520.0, ymax=5412480.0)
```

**Comment by vincentsarago** [comment]
name | CRS | inverted | NorthingEasting | latlong
--- | ---  | --- | ---      | ---
LINZAntarticaMapTilegrid | EPSG:5482 | True | True | False
EuropeanETRS89_LAEAQuad | EPSG:3035 | True | True | False
CanadianNAD83_LCC | EPSG:3978 | False | False | False
UPSArcticWGS84Quad | EPSG:5041 | False | False | False
NZTM2000 | EPSG:2193 | True | True | False
UTM31WGS84Quad | EPSG:32631 | False | False | False
UPSAntarcticWGS84Quad | EPSG:5042 | False | False | False
WorldMercatorWGS84Quad | EPSG:3395 | False | False | False
WorldCRS84Quad | EPSG:4326 | False | False | False
WebMercatorQuad | EPSG:3857 | False | False | False


There is something odd about the `EuropeanETRS89_LAEAQuad` and `LINZAntarticaMapTilegrid ` TMS documents. While GDAL (and proj) tells we should invert the axis, the coordinates are stored in x,y order 🤔

**Comment by palmerj** [comment]
I provided the scales, origin and EPSG ID, Even made the JSON file. Note we have just been discussing this projection over in https://github.com/OSGeo/PROJ/issues/2254. The projection axis is swapped and has a very usually definition. See the comment from Even:
> "easting (E)",north, means that this is an axis meant at representing easting values of the projection, but as you are centered at the south pole, all directions starting from it go to the north.

> While GDAL (and proj) tells we should invert the axis, the coordinates are stored in x,y order

What do you mean by this?

**Comment by vincentsarago** [comment]
Excellent, 

> What do you mean by this?

```
    def _invert_axis(self) -> bool:
        """Check if CRS has inverted AXIS (lat,lon) or (lon,lat)."""
        return (
            epsg_treats_as_latlong(self.crs)
            or epsg_treats_as_northingeasting(self.crs)
        )
```
I mean ☝️ return `TRUE` for `http://www.opengis.net/def/crs/EPSG/0/5482` CRS, thus when creating the tiles, I'm swapping the coordinates (like for the `NZTM2000` TMS) but the result won't be good because when I check the TMS document (json) coordinates, the origin
```
"topLeftCorner": [
    6918457.73,
    -918457.73
],
```
is stored as X, Y 

e.g for `NZTM2000`
```
"topLeftCorner": [
    10000000,
    -1000000
],
```
the origin is stored as Y,X

**Comment by palmerj** [comment]
> "topLeftCorner": [
    6918457.73,
    -918457.73
],
is stored as X, Y

This is actually Y,X. So it looks good.

File: setup.py (hunk 4 above)
Comment (vincentsarago) [reply setup.py:8]: will in fact be 1.2.0

**Comment by vincentsarago** [comment]
Update, this is still blocked on rasterio side, but Sean might move forward by adding the need API in rasterio sooner than in 1.2 https://github.com/mapbox/rasterio/pull/1943#issuecomment-700183778 🎉

**Pull request merged**